src/api.js

```diff
@@ -464,56 +511,104 @@ var remoteAuthentication = function(options, entry) {
+            } else {
+              subexpressions.push(_expandExpressionTemplate(scope));
             }
           });
+          return {[key]: subexpressions};
+        };
+
+        var scopeExpression = _expandExpressionTemplate(entry.scopes);
+
+        if (hasConditional && allowLater) {
+          debug('Conditional provided in scopes. Deferring authorization.');
+          return true;
+        }
+
+        if (missing.length) {
+          if (allowLater) {
+            debug(`Not all parameters supplied yet. Deferring authorization due to missing parameters: ${req.missing}`);
+            return true;
+          }
+          return res.reportInternalError('InternalServerError', [
+            'Not all parameters were supplied to a scope check.',
+            'The call to req.authorize was not allowed to be partially',
+            'applied. Missing parameters: {{missing}}',
+          ].join('\n'), {missing: req.missing});
         }
 
         // Test that we have scope intersection, and hence, is authorized
-        var retval = scopes.scopeMatch(result.scopes, scopesets);
+        var retval = scopes.satisfiesExpression(result.scopes, scopeExpression);
+        req.hasAuthed = true;
         if (retval) {
           // TODO: log this in a structured format when structured logging is
           // available https://bugzilla.mozilla.org/show_bug.cgi?id=1307271
-          authLog(`Authorized ${clientId} for ${req.method} access to ${req.originalUrl}`);
+          authLog(`Authorized ${await req.clientId()} for ${req.method} access to ${req.originalUrl}`);
         }
         if (!retval && !noReply) {
+          let missing = scopes.removeGivenScopes(result.scopes, scopeExpression);
           res.reportError('InsufficientScopes', [
-            'You do not have sufficient scopes. This request requires you',
-            'to have one of the following sets of scopes:',
-            '{{scopesets}}',
+            'You do not have sufficient scopes. You are missing the following scopes:',
+            '',
+            '{{missing}}',
             '',
             'You only have the scopes:',
+            '',
             '{{scopes}}',
             '',
-            'In other words you are missing scopes from one of the options:',
-          ].concat(scopesets.map((set, index) => {
-            let missing = set.filter(scope => {
-              return !scopes.scopeMatch(result.scopes, [[scope]]);
-            });
-            return ' * Option ' + index + ':\n    - "' +
-                   missing.join('", and\n    - "') + '"';
-          })).join('\n'),  {scopesets, scopes: result.scopes});
+            'This request requires you to satisfy this scope expression:',
+            '',
+            '{{scopeExpression}}',
+          ]
+            .join('\n'),  {missing, scopeExpression, scopes: result.scopes});
         }
         return retval;
       };
 
+      req.hasAuthed = true; // No need to check auth unless there are scopes
```

Comment:
It took me a while to figure out that this applies only when `!entry.scopes`, since `req.authorize` will set it back to false.  It also means that for the briefest of moments, `req.hasAuthed` is true even when the req has not been authed.  Maybe expand the conditional below a little?

```
req.hasAuthed = false;

if (!entry.scopes) {
  req.hasAuthed = true;  // No need to check auth if there are no scopes
  next();
} else if (await req.authorize(req.params, {allowLater: true})) {
  next();
}
```

Also, I'm confused how this function appears to only call `next()` in some circumstances.. but that's not new with this PR.